File: GO/bst.go
package main

import (
	"fmt"
	"time"
)

func Max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

type Node struct {
	key    int
	left   *Node
	right  *Node
	height int
}

type BST struct {
	root *Node
}

type AVL struct {
	root *Node
}

// Node

func (node *Node) insert(data int) {
	if data <= node.key {
		if node.left == nil {
			node.left = &Node{key: data}
		} else {
			node.left.insert(data)
		}
	} else {
		if node.right == nil {
			node.right = &Node{key: data}
		} else {
			node.right.insert(data)
		}
	}
}

// Tree

func (tree *BST) insert(data int) {
	if tree.root == nil {
		tree.root = &Node{key: data}
	} else {
		tree.root.insert(data)
	}
}

func NewNode(key int) *Node {
	return &Node{
		key:    key,
		left:   nil,
		right:  nil,
		height: 1,
	}
}

func main() {
	for i := 10000; i <= 100000; i += 10000 {
		var result1 float64 = 0
		var result2 float64 = 0
		Arr := make([]int, i)
		for j := 0; j < i; j++ {
			Arr[j] = i - j - 1
		}
		bst := NewBST()
		bst.InsertArray(Arr)
		avl := NewAVL()
		avl.InsertArray(Arr)
		for k := 0; k < 10000; k++ {
			start := time.Now()
			bst.root.InOrder()
			duration := time.Since(start)
			result1 += float64(duration.Nanoseconds()) / 1000000.00
			start = time.Now()
			avl.root.InOrder()
			duration = time.Since(start)
			result2 += float64(duration.Nanoseconds()) / 1000000.00
		}
		result1 /= 10000.00
		result2 /= 10000.00
		fmt.Printf("%v elements: BST - %v ms, AVL - %v ms\n", i, result1, result2)
	}
}

File: C/bst.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct BST BST;

struct BST {
    int key;
    BST* left;
    BST* right;
    int height;
};

BST* create(int key) {
    BST bst = {key, NULL, NULL, 1};
    return &bst;
}

void insert(BST* tree, int key) {
    if (!tree) {
        tree = create(key);
        return;
    }
    if (tree->key == key) {
        return;
    }
    if (tree->key < key) {
        insert(tree->right, key);
    } else {
        insert(tree->left, key);
    }
    tree->height = 1 + fmax(getHeight(tree->left), getHeight(tree->right));
}

/*
void insertArr(BST* tree, int* arr, int len) {
    for (int i = 0; i < len; insert(tree, arr[i++]));
} */

void delete(BST* tree, int key) {
    if (!tree) {
        return;
    }
    if (key < tree->key) {
        delete(tree->left, key);
    } else if (key > tree->key) {
        delete(tree->right, key);
    } else {
        if (!tree->left) {
            tree = tree->right;
            return;
        } else {
            tree = tree->left;
            return;
        }
        BST* tmp;
    }
}

int getHeight(BST* node) {
    if (!node) {
        return 0;
    }
    return node->height;
}

BST* minValue(BST* node) {
    BST* current = node;
}

int main(void) {
    printf("Hello world\n");
    BST root;
    root.key = 1;
}